File: 202-happy-number/202-happy-number.py
class Solution:
    def isHappy(self, n: int) -> bool:
        def sum_square(num):
            num_sum = 0
            while num!=0:
                num_sum += (num%10)**2
                num = num//10
            return num_sum
        visited = set()
        while n != 1:
            if n in visited:
                return False
            else:
                visited.add(n)
                n = sum_square(n)
        return True
        

File: 202-happy-number/202-happy-number.java
class Solution {
    public boolean isHappy(int n) {
        int slowNum = n;
        int fastNum = n;
        do {
            slowNum = getNextNumber(slowNum);
            fastNum = getNextNumber(getNextNumber(fastNum));
        } while(fastNum!=1 && slowNum!=fastNum);
        return fastNum==1;
    }
    private int getNextNumber(int n){
        int prod = 0;
        while(n>0){
            prod += (n%10)*(n%10);
            n /= 10;
        }
        return prod;
    }
}